dozer-lambda/src/js/worker/mod.rs

```diff
@@ -0,0 +1,285 @@
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    future::poll_fn,
+    num::NonZeroI32,
+    ops::ControlFlow,
+    rc::Rc,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use deno_runtime::{
+    deno_core::{error::AnyError, JsRuntime, ModuleSpecifier},
+    deno_napi::v8,
+    permissions::PermissionsContainer,
+    worker::{MainWorker, WorkerOptions},
+};
+use dozer_log::{
+    errors::ReaderBuilderError,
+    tokio::{
+        runtime::Runtime,
+        sync::{
+            mpsc::{channel, error::TryRecvError, Receiver, Sender},
+            oneshot, Mutex,
+        },
+        task::{JoinError, JoinHandle, LocalSet},
+    },
+};
+use dozer_types::{
+    log::{error, info},
+    models::lambda_config::JavaScriptLambda,
+    thiserror,
+    tracing::trace,
+    types::{Field, Operation},
+};
+
+use super::trigger::Trigger;
+
+#[derive(Debug, Clone)]
+pub struct Worker {
+    work_sender: Sender<Work>,
+    handle: Arc<Mutex<Option<JoinHandle<()>>>>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("failed to initialize JavaScript lambda runtime: {0:?}")]
+    InitPanic(#[source] Option<JoinError>),
+    #[error("failed to canonicalize path {0}: {1}")]
+    CanonicalizePath(String, #[source] std::io::Error),
+    #[error("failed to load lambda module {0}: {1}")]
+    LoadLambdaModule(String, #[source] AnyError),
+    #[error("failed to evaluate lambda module {0}: {1}")]
+    EvaluateLambdaModule(String, #[source] AnyError),
+    #[error("failed to get namespace of lambda module {0}: {1}")]
+    GetLambdaModuleNamespace(String, #[source] AnyError),
+    #[error("lambda module {0} has no default export")]
+    LambdaModuleNoDefaultExport(String),
+    #[error("lambda module {0} default export is not a function: {1}")]
+    LambdaModuleDefaultExportNotFunction(String, #[source] v8::DataError),
+    #[error("failed to register lambda {0} for endpoint {1}: {2}")]
+    RegisterLambda(String, String, #[source] ReaderBuilderError),
+}
+
+impl Worker {
+    pub async fn new(
+        runtime: Arc<Runtime>,
+        trigger: Arc<Mutex<Trigger>>,
+        lambda_modules: Vec<JavaScriptLambda>,
+    ) -> Result<Self, Error> {
+        let (init_sender, init_receiver) = oneshot::channel();
+        let (work_sender, work_receiver) = channel(1);
+        let runtime_clone = runtime.clone();
+        let handle = runtime.spawn_blocking(move || {
+            let lambdas = Rc::new(RefCell::new(HashMap::new()));
+            let mut main_worker = MainWorker::bootstrap_from_options(
```

Comment:
I think we want to initialize a `JsRuntime` ourselves here. The `MainWorker` provides many deno extensions, a lot of which we don't want if we want to properly sandbox lambdas (such as the fs events extension). See, for example, supabases edge runtime: https://github.com/supabase/edge-runtime/blob/b2092c646ca83a1a0810bce5d1476656e50374a4/crates/base/src/deno_runtime.rs#L206

We can always add more extensions if we deem them safe and useful, but removing extensions will break users' code.

Reply:
The main reason I'm using `MainWorker` is for [`MainWorker::evaluate_module`](https://docs.rs/deno_runtime/latest/src/deno_runtime/worker.rs.html#442-464). Don't feel like copying that snippet around.

How about we use `PermissionsContainer` to control permissions?

I also believe `alow_all` is suitable here, as user should be able to do whatever they want when self-hosting. We can blacklist `fs` in cloud.

Reply:
If we want to be able to await the results of lambda's, for the sake of durability, we can't allow using the `Worker` API or subprocesses. Besides that, I think DX is better if the capabilities when running locally are the same as on cloud